clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/JsonHubProtocol.java

```diff
@@ -32,41 +35,105 @@ public TransferFormat getTransferFormat() {
+                            reader.beginArray();
+                            List<Class<?>> types = binder.GetParameterTypes(target);
+                            if (types != null && types.size() >= 1) {
+                                arguments = new ArrayList<>();
+                                for (int i = 0; i < types.size(); i++) {
+                                    arguments.add(gson.fromJson(reader, types.get(i)));
+                                }
+                            }
+                            reader.endArray();
+                        } else {
+                            argumentsToken = (JsonArray)jsonParser.parse(reader);
+                        }
+                        break;
+                    case "headers":
+                        throw new HubException("Headers not implemented yet.");
+                    default:
+                        // Skip unknown property, allows new clients to still work with old protocols
+                        reader.skipValue();
+                        break;
+                }
+            } while (reader.hasNext());
+
+            reader.endObject();
+            reader.close();
 
-            JsonObject jsonMessage = jsonParser.parse(splitMessage).getAsJsonObject();
-            HubMessageType messageType = HubMessageType.values()[jsonMessage.get("type").getAsInt() -1];
             switch (messageType) {
                 case INVOCATION:
-                    //Invocation Message
-                    String target = jsonMessage.get("target").getAsString();
-                    JsonElement args = jsonMessage.get("arguments");
-                    hubMessages.add(new InvocationMessage(target, new Object[] {args}));
+                    if (argumentsToken != null) {
+                        List<Class<?>> types = binder.GetParameterTypes(target);
+                        if (types != null && types.size() >= 1) {
+                            arguments = new ArrayList<>();
+                            for (int i = 0; i < types.size(); i++) {
+                                arguments.add(gson.fromJson(argumentsToken.get(i), types.get(i)));
+                            }
+                        }
+                    }
+                    if (arguments == null) {
+                        hubMessages.add(new InvocationMessage(target, new Object[0]));
+                    } else {
+                        hubMessages.add(new InvocationMessage(target, arguments.toArray()));
+                    }
                     break;
+                case STREAM_INVOCATION:
                 case STREAM_ITEM:
                 case COMPLETION:
-                case STREAM_INVOCATION:
                 case CANCEL_INVOCATION:
                     throw new UnsupportedOperationException(String.format("The message type %s is not supported yet.", messageType));
                 case PING:
-                    //Ping
                     hubMessages.add(new PingMessage());
```

Comment:
While you're here can you quickly bash up a singleton for this? [Looks like](https://www.journaldev.com/1377/java-singleton-design-pattern-best-practices-examples) the java pattern would be: 

```java
class PingMessage
{
    private static final PingMessage instance = new PingMessage();
    
    private PingMessage()
    {
    }

    public static PingMessage getInstance(){
        return instance;
    }
}

Reply:
I can do this as part of API review changes. No need to add it here

Reply:
As long as it's tracked somewhere